compiler/rustc_ast_pretty/src/pprust/state/expr.rs

```diff
@@ -117,38 +117,46 @@ impl<'a> State<'a> {
         } else {
             self.print_path(path, true, 0);
         }
+        self.nbsp();
         self.word("{");
-        self.commasep_cmnt(
-            Consistent,
-            fields,
-            |s, field| {
-                s.print_outer_attributes(&field.attrs);
-                s.ibox(INDENT_UNIT);
-                if !field.is_shorthand {
-                    s.print_ident(field.ident);
-                    s.word_space(":");
-                }
-                s.print_expr(&field.expr);
-                s.end();
-            },
-            |f| f.span,
-        );
-        match rest {
-            ast::StructRest::Base(_) | ast::StructRest::Rest(_) => {
-                self.ibox(INDENT_UNIT);
-                if !fields.is_empty() {
-                    self.word(",");
-                    self.space();
-                }
-                self.word("..");
-                if let ast::StructRest::Base(ref expr) = *rest {
-                    self.print_expr(expr);
-                }
-                self.end();
+        let has_rest = match rest {
+            ast::StructRest::Base(_) | ast::StructRest::Rest(_) => true,
+            ast::StructRest::None => false,
+        };
+        if fields.is_empty() && !has_rest {
+            self.word("}");
+            return;
+        }
+        self.cbox(0);
+        for field in fields.iter().delimited() {
+            self.maybe_print_comment(field.span.hi());
+            self.print_outer_attributes(&field.attrs);
+            if field.is_first {
+                self.space_if_not_bol();
+            }
+            if !field.is_shorthand {
+                self.print_ident(field.ident);
+                self.word_nbsp(":");
+            }
+            self.print_expr(&field.expr);
+            if !field.is_last || has_rest {
+                self.word_space(",");
+            } else {
+                self.trailing_comma();
             }
-            ast::StructRest::None if !fields.is_empty() => self.word(","),
-            _ => {}
         }
+        if has_rest {
+            if fields.is_empty() {
+                self.space();
+            }
+            self.word("..");
+            if let ast::StructRest::Base(expr) = rest {
+                self.print_expr(expr);
+            }
+            self.space();
+        }
+        self.offset(-INDENT_UNIT);
```

Comment:
Could you explain what this negative offset does? I have trouble understanding how this code manages to find the correct indentation.

Reply:
The top of https://github.com/rust-lang/rust/blob/1.58.1/compiler/rustc_ast_pretty/src/pp.rs has an overview of the printing algorithm. Very briefly, when you make a sequence of <kbd>cbox</kbd> <kbd>string</kbd> <kbd>break</kbd> <kbd>string</kbd> <kbd>break</kbd> <kbd>string</kbd> <kbd>break</kbd> <kbd>string</kbd> <kbd>end</kbd> calls on the printer, either all of those breaks will turn into linebreaks, or none of them will turn into linebreaks. If they turn into linebreaks, the indentation at each one is the sum of the indentation of all the surrounding boxes.

So if we have:

<pre>
<kbd>ibox(4)</kbd><kbd>cbox(0)</kbd>Struct {<kbd>break</kbd>a: A<kbd>break</kbd>b: B<kbd>break</kbd>}<kbd>end</kbd><kbd>end</kbd>
</pre>

and it does not fit on the line, then you'd expect to get:

```console
Struct {
    a: A
    b: B
    }
```

That's why every <kbd>break</kbd> also holds its own offset which is relative to the sum of the surrounding boxes' indents, so that we can set -4 on the last break and get:

```console
Struct {
    a: A
    b: B
}
```

Reply:
Thank you for the detailed explanation. I'm surprised such negative offset facility was not used before, but that's maybe for another refactoring.